File: src/aigpt/commands/submodules.py
"""Submodule management commands for ai.gpt."""

from pathlib import Path
from typing import Dict, List, Optional, Tuple
import subprocess
import json

import typer
from rich.console import Console
from rich.panel import Panel
from rich.table import Table

from ..docs.config import get_ai_root, load_docs_config
from ..docs.git_utils import (
    check_git_repository, 
    get_git_branch, 
    get_git_remote_url
)
from ..docs.utils import run_command

console = Console()
submodules_app = typer.Typer(help="Submodule management for AI ecosystem")


def get_submodules_from_gitmodules(repo_path: Path) -> Dict[str, str]:
    """Parse .gitmodules file to get submodule information."""
    gitmodules_path = repo_path / ".gitmodules"
    if not gitmodules_path.exists():
        return {}
    
    submodules = {}
    current_name = None
    
    with open(gitmodules_path, 'r') as f:
        for line in f:
            line = line.strip()
            if line.startswith('[submodule "') and line.endswith('"]'):
                current_name = line[12:-2]  # Extract module name
            elif line.startswith('path = ') and current_name:
                path = line[7:]  # Extract path
                submodules[current_name] = path
                current_name = None
    
    return submodules


def get_branch_for_module(config, module_name: str) -> str:
    """Get target branch for a module from ai.json."""
    project_info = config.get_project_info(module_name)
    if project_info and project_info.branch:
        return project_info.branch
    return "main"  # Default branch


@submodules_app.command("list")
def list_submodules(
    dir: Optional[Path] = typer.Option(None, "--dir", "-d", help="AI ecosystem root directory"),
    verbose: bool = typer.Option(False, "--verbose", "-v", help="Show detailed information")
):
    """List all submodules and their status."""
    try:
        config = load_docs_config(dir)
        ai_root = get_ai_root(dir)
        
        if not check_git_repository(ai_root):
            console.print("[red]Error: Not a git repository[/red]")
            raise typer.Abort()
        
        submodules = get_submodules_from_gitmodules(ai_root)
        
        if not submodules:
            console.print("[yellow]No submodules found[/yellow]")
            return
        
        table = Table(title="Submodules Status")
        table.add_column("Module", style="cyan")
        table.add_column("Path", style="blue")
        table.add_column("Branch", style="green")
        table.add_column("Status", style="yellow")
        
        for module_name, module_path in submodules.items():
            full_path = ai_root / module_path
            
            if not full_path.exists():
                status = "❌ Missing"
                branch = "N/A"
            else:
                branch = get_git_branch(full_path) or "detached"
                
                # Check if submodule is up to date
                returncode, stdout, stderr = run_command(
                    ["git", "submodule", "status", module_path],
                    cwd=ai_root
                )
                
                if returncode == 0 and stdout:
                    status_char = stdout[0] if stdout else ' '
                    if status_char == ' ':
                        status = "✅ Clean"
                    elif status_char == '+':
                        status = "📝 Modified"
                    elif status_char == '-':
                        status = "❌ Not initialized"
                    elif status_char == 'U':
                        status = "⚠️ Conflicts"
                    else:
                        status = "❓ Unknown"
                else:
                    status = "❓ Unknown"
            
            target_branch = get_branch_for_module(config, module_name)
            branch_display = f"{branch}"
            if branch != target_branch:
                branch_display += f" (target: {target_branch})"
            
            table.add_row(module_name, module_path, branch_display, status)
        
        console.print(table)
        
        if verbose:
            console.print(f"\n[dim]Total submodules: {len(submodules)}[/dim]")
            console.print(f"[dim]Repository root: {ai_root}[/dim]")
        
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
        raise typer.Abort()


@submodules_app.command("update")
def update_submodules(
    module: Optional[str] = typer.Option(None, "--module", "-m", help="Update specific submodule"),
    all: bool = typer.Option(False, "--all", "-a", help="Update all submodules"),
    dir: Optional[Path] = typer.Option(None, "--dir", "-d", help="AI ecosystem root directory"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Show what would be done"),
    auto_commit: bool = typer.Option(False, "--auto-commit", help="Auto-commit changes"),
    verbose: bool = typer.Option(False, "--verbose", "-v", help="Show detailed output")
):
    """Update submodules to latest commits."""
    if not module and not all:
        console.print("[red]Error: Either --module or --all is required[/red]")
        raise typer.Abort()
    
    if module and all:
        console.print("[red]Error: Cannot use both --module and --all[/red]")
        raise typer.Abort()
    
    try:
        config = load_docs_config(dir)
        ai_root = get_ai_root(dir)
        
        if not check_git_repository(ai_root):
            console.print("[red]Error: Not a git repository[/red]")
            raise typer.Abort()
        
        submodules = get_submodules_from_gitmodules(ai_root)
        
        if not submodules:
            console.print("[yellow]No submodules found[/yellow]")
            return
        
        # Determine which modules to update
        if all:
            modules_to_update = list(submodules.keys())
        else:
            if module not in submodules:
                console.print(f"[red]Error: Submodule '{module}' not found[/red]")
                console.print(f"Available modules: {', '.join(submodules.keys())}")
                raise typer.Abort()
            modules_to_update = [module]
        
        if dry_run:
            console.print("[yellow]🔍 DRY RUN MODE - No changes will be made[/yellow]")
        
        console.print(f"[cyan]Updating {len(modules_to_update)} submodule(s)...[/cyan]")
        
        updated_modules = []
        
        for module_name in modules_to_update:
            module_path = submodules[module_name]
            full_path = ai_root / module_path
            target_branch = get_branch_for_module(config, module_name)
            
            console.print(f"\n[blue]📦 Processing: {module_name}[/blue]")
            
            if not full_path.exists():
                console.print(f"[red]❌ Module directory not found: {module_path}[/red]")
                continue
            
            # Get current commit
            current_commit = None
            returncode, stdout, stderr = run_command(
                ["git", "rev-parse", "HEAD"],
                cwd=full_path
            )
            if returncode == 0:
                current_commit = stdout.strip()[:8]
            
            if dry_run:
                console.print(f"[yellow]🔍 Would update {module_name} to branch {target_branch}[/yellow]")
                if current_commit:
                    console.print(f"[dim]Current: {current_commit}[/dim]")
                continue
            
            # Fetch latest changes
            console.print(f"[dim]Fetching latest changes...[/dim]")
            returncode, stdout, stderr = run_command(
                ["git", "fetch", "origin"],
                cwd=full_path
            )
            
            if returncode != 0:
                console.print(f"[red]❌ Failed to fetch: {stderr}[/red]")
                continue
            
            # Check if update is needed
            returncode, stdout, stderr = run_command(
                ["git", "rev-parse", f"origin/{target_branch}"],
                cwd=full_path
            )
            
            if returncode != 0:
                console.print(f"[red]❌ Branch {target_branch} not found on remote[/red]")
                continue
            
            latest_commit = stdout.strip()[:8]
            
            if current_commit == latest_commit:
                console.print(f"[green]✅ Already up to date[/green]")
                continue
            
            # Switch to target branch and pull
            console.print(f"[dim]Switching to branch {target_branch}...[/dim]")
            returncode, stdout, stderr = run_command(
                ["git", "checkout", target_branch],
                cwd=full_path
            )
            
            if returncode != 0:
                console.print(f"[red]❌ Failed to checkout {target_branch}: {stderr}[/red]")
                continue
            
            returncode, stdout, stderr = run_command(
                ["git", "pull", "origin", target_branch],
                cwd=full_path
            )
            
            if returncode != 0:
                console.print(f"[red]❌ Failed to pull: {stderr}[/red]")
                continue
            
            # Get new commit
            returncode, stdout, stderr = run_command(
                ["git", "rev-parse", "HEAD"],
                cwd=full_path
            )
            new_commit = stdout.strip()[:8] if returncode == 0 else "unknown"
            
            # Stage the submodule update
            returncode, stdout, stderr = run_command(
                ["git", "add", module_path],
                cwd=ai_root
            )
            
            console.print(f"[green]✅ Updated {module_name} ({current_commit} → {new_commit})[/green]")
            updated_modules.append((module_name, current_commit, new_commit))
        
        # Summary
        if updated_modules:
            console.print(f"\n[green]🎉 Successfully updated {len(updated_modules)} module(s)[/green]")
            
            if verbose:
                for module_name, old_commit, new_commit in updated_modules:
                    console.print(f"  • {module_name}: {old_commit} → {new_commit}")
            
            if auto_commit and not dry_run:
                console.print("[blue]💾 Auto-committing changes...[/blue]")
                commit_message = f"Update submodules\n\n📦 Updated modules: {len(updated_modules)}\n"
                for module_name, old_commit, new_commit in updated_modules:
                    commit_message += f"- {module_name}: {old_commit} → {new_commit}\n"
                commit_message += "\n🤖 Generated with ai.gpt submodules update"
                
                returncode, stdout, stderr = run_command(
                    ["git", "commit", "-m", commit_message],
                    cwd=ai_root
                )
                
                if returncode == 0:
                    console.print("[green]✅ Changes committed successfully[/green]")
                else:
                    console.print(f"[red]❌ Failed to commit: {stderr}[/red]")
            elif not dry_run:
                console.print("[yellow]💾 Changes staged but not committed[/yellow]")
                console.print("Run with --auto-commit to commit automatically")
        elif not dry_run:
            console.print("[yellow]No modules needed updating[/yellow]")
        
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
        if verbose:
            console.print_exception()
        raise typer.Abort()


# Export the submodules app
__all__ = ["submodules_app"]

File: aigpt-rs/src/submodules.rs
use std::collections::HashMap;
use std::path::PathBuf;
use anyhow::{Result, Context};
use colored::*;
use serde::{Deserialize, Serialize};

use crate::config::Config;

pub async fn handle_submodules(
    action: String,
    module: Option<String>,
    all: bool,
    dry_run: bool,
    auto_commit: bool,
    verbose: bool,
    data_dir: Option<PathBuf>,
) -> Result<()> {
    let config = Config::new(data_dir)?;
    let mut submodule_manager = SubmoduleManager::new(config);
    
    match action.as_str() {
        "list" => {
            submodule_manager.list_submodules(verbose).await?;
        }
        "update" => {
            submodule_manager.update_submodules(module, all, dry_run, auto_commit, verbose).await?;
        }
        "status" => {
            submodule_manager.show_submodule_status().await?;
        }
        _ => {
            return Err(anyhow::anyhow!("Unknown submodule action: {}", action));
        }
    }
    
    Ok(())
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmoduleInfo {
    pub name: String,
    pub path: String,
    pub branch: String,
    pub current_commit: Option<String>,
    pub target_commit: Option<String>,
    pub status: String,
}

impl Default for SubmoduleInfo {
    fn default() -> Self {
        SubmoduleInfo {
            name: String::new(),
            path: String::new(),
            branch: "main".to_string(),
            current_commit: None,
            target_commit: None,
            status: "unknown".to_string(),
        }
    }
}

pub struct SubmoduleManager {
    config: Config,
    ai_root: PathBuf,
    submodules: HashMap<String, SubmoduleInfo>,
}

impl SubmoduleManager {
    pub fn new(config: Config) -> Self {
        let ai_root = dirs::home_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join("ai")
            .join("ai");
        
        SubmoduleManager {
            config,
            ai_root,
            submodules: HashMap::new(),
        }
    }
    
    pub async fn list_submodules(&mut self, verbose: bool) -> Result<()> {
        println!("{}", "📋 Submodules Status".cyan().bold());
        println!();
        
        let submodules = self.parse_gitmodules()?;
        
        if submodules.is_empty() {
            println!("{}", "No submodules found".yellow());
            return Ok(());
        }
        
        // Display submodules in a table format
        println!("{:<15} {:<25} {:<15} {}", 
                "Module".cyan().bold(), 
                "Path".cyan().bold(), 
                "Branch".cyan().bold(), 
                "Status".cyan().bold());
        println!("{}", "-".repeat(80));
        
        for (module_name, module_info) in &submodules {
            let status_color = match module_info.status.as_str() {
                "clean" => module_info.status.green(),
                "modified" => module_info.status.yellow(),
                "missing" => module_info.status.red(),
                "conflicts" => module_info.status.red(),
                _ => module_info.status.normal(),
            };
            
            println!("{:<15} {:<25} {:<15} {}", 
                    module_name.blue(),
                    module_info.path,
                    module_info.branch.green(),
                    status_color);
        }
        
        println!();
        
        if verbose {
            println!("Total submodules: {}", submodules.len().to_string().cyan());
            println!("Repository root: {}", self.ai_root.display().to_string().blue());
        }
        
        Ok(())
    }
    
    pub async fn update_submodules(
        &mut self, 
        module: Option<String>, 
        all: bool, 
        dry_run: bool, 
        auto_commit: bool, 
        verbose: bool
    ) -> Result<()> {
        if !module.is_some() && !all {
            return Err(anyhow::anyhow!("Either --module or --all is required"));
        }
        
        if module.is_some() && all {
            return Err(anyhow::anyhow!("Cannot use both --module and --all"));
        }
        
        let submodules = self.parse_gitmodules()?;
        
        if submodules.is_empty() {
            println!("{}", "No submodules found".yellow());
            return Ok(());
        }
        
        // Determine which modules to update
        let modules_to_update: Vec<String> = if all {
            submodules.keys().cloned().collect()
        } else if let Some(module_name) = module {
            if !submodules.contains_key(&module_name) {
                return Err(anyhow::anyhow!(
                    "Submodule '{}' not found. Available modules: {}", 
                    module_name, 
                    submodules.keys().cloned().collect::<Vec<_>>().join(", ")
                ));
            }
            vec![module_name]
        } else {
            vec![]
        };
        
        if dry_run {
            println!("{}", "🔍 DRY RUN MODE - No changes will be made".yellow().bold());
        }
        
        println!("{}", format!("🔄 Updating {} submodule(s)...", modules_to_update.len()).cyan().bold());
        
        let mut updated_modules = Vec::new();
        
        for module_name in modules_to_update {
            if let Some(module_info) = submodules.get(&module_name) {
                println!("\n{}", format!("📦 Processing: {}", module_name).blue().bold());
                
                let module_path = PathBuf::from(&module_info.path);
                let full_path = self.ai_root.join(&module_path);
                
                if !full_path.exists() {
                    println!("{}", format!("❌ Module directory not found: {}", module_info.path).red());
                    continue;
                }
                
                // Get current commit
                let current_commit = self.get_current_commit(&full_path)?;
                
                if dry_run {
                    println!("{}", format!("🔍 Would update {} to branch {}", module_name, module_info.branch).yellow());
                    if let Some(ref commit) = current_commit {
                        println!("{}", format!("Current: {}", commit).dimmed());
                    }
                    continue;
                }
                
                // Perform update
                if let Err(e) = self.update_single_module(&module_name, &module_info, &full_path).await {
                    println!("{}", format!("❌ Failed to update {}: {}", module_name, e).red());
                    continue;
                }
                
                // Get new commit
                let new_commit = self.get_current_commit(&full_path)?;
                
                if current_commit != new_commit {
                    println!("{}", format!("✅ Updated {} ({:?} → {:?})", 
                           module_name, 
                           current_commit.as_deref().unwrap_or("unknown"),
                           new_commit.as_deref().unwrap_or("unknown")).green());
                    updated_modules.push((module_name.clone(), current_commit, new_commit));
                } else {
                    println!("{}", "✅ Already up to date".green());
                }
            }
        }
        
        // Summary
        if !updated_modules.is_empty() {
            println!("\n{}", format!("🎉 Successfully updated {} module(s)", updated_modules.len()).green().bold());
            
            if verbose {
                for (module_name, old_commit, new_commit) in &updated_modules {
                    println!("  • {}: {:?} → {:?}", 
                           module_name,
                           old_commit.as_deref().unwrap_or("unknown"),
                           new_commit.as_deref().unwrap_or("unknown"));
                }
            }
            
            if auto_commit && !dry_run {
                self.auto_commit_changes(&updated_modules).await?;
            } else if !dry_run {
                println!("{}", "💾 Changes staged but not committed".yellow());
                println!("Run with --auto-commit to commit automatically");
            }
        } else if !dry_run {
            println!("{}", "No modules needed updating".yellow());
        }
        
        Ok(())
    }
    
    pub async fn show_submodule_status(&self) -> Result<()> {
        println!("{}", "📊 Submodule Status Overview".cyan().bold());
        println!();
        
        let submodules = self.parse_gitmodules()?;
        let mut total_modules = 0;
        let mut clean_modules = 0;
        let mut modified_modules = 0;
        let mut missing_modules = 0;
        
        for (module_name, module_info) in submodules {
            let module_path = self.ai_root.join(&module_info.path);
            
            if module_path.exists() {
                total_modules += 1;
                match module_info.status.as_str() {
                    "clean" => clean_modules += 1,
                    "modified" => modified_modules += 1,
                    _ => {}
                }
            } else {
                missing_modules += 1;
            }
            
            println!("{}: {}", 
                    module_name.blue(), 
                    if module_path.exists() {
                        module_info.status.green()
                    } else {
                        "missing".red()
                    });
        }
        
        println!();
        println!("Summary: {} total, {} clean, {} modified, {} missing", 
                total_modules.to_string().cyan(), 
                clean_modules.to_string().green(),
                modified_modules.to_string().yellow(),
                missing_modules.to_string().red());
        
        Ok(())
    }
    
    fn parse_gitmodules(&self) -> Result<HashMap<String, SubmoduleInfo>> {
        let gitmodules_path = self.ai_root.join(".gitmodules");
        
        if !gitmodules_path.exists() {
            return Ok(HashMap::new());
        }
        
        let content = std::fs::read_to_string(&gitmodules_path)
            .with_context(|| format!("Failed to read .gitmodules file: {}", gitmodules_path.display()))?;
        
        let mut submodules = HashMap::new();
        let mut current_name: Option<String> = None;
        let mut current_path: Option<String> = None;
        
        for line in content.lines() {
            let line = line.trim();
            
            if line.starts_with("[submodule \"") && line.ends_with("\"]") {
                // Save previous submodule if complete
                if let (Some(name), Some(path)) = (current_name.take(), current_path.take()) {
                    let mut info = SubmoduleInfo::default();
                    info.name = name.clone();
                    info.path = path;
                    info.branch = self.get_target_branch(&name);
                    info.status = self.get_submodule_status(&name, &info.path)?;
                    submodules.insert(name, info);
                }
                
                // Extract new submodule name
                current_name = Some(line[12..line.len()-2].to_string());
            } else if line.starts_with("path = ") {
                current_path = Some(line[7..].to_string());
            }
        }
        
        // Save last submodule
        if let (Some(name), Some(path)) = (current_name, current_path) {
            let mut info = SubmoduleInfo::default();
            info.name = name.clone();
            info.path = path;
            info.branch = self.get_target_branch(&name);
            info.status = self.get_submodule_status(&name, &info.path)?;
            submodules.insert(name, info);
        }
        
        Ok(submodules)
    }
    
    fn get_target_branch(&self, module_name: &str) -> String {
        // Try to get from ai.json configuration
        match module_name {
            "verse" => "main".to_string(),
            "card" => "main".to_string(),
            "bot" => "main".to_string(),
            _ => "main".to_string(),
        }
    }
    
    fn get_submodule_status(&self, _module_name: &str, module_path: &str) -> Result<String> {
        let full_path = self.ai_root.join(module_path);
        
        if !full_path.exists() {
            return Ok("missing".to_string());
        }
        
        // Check git status
        let output = std::process::Command::new("git")
            .args(&["submodule", "status", module_path])
            .current_dir(&self.ai_root)
            .output();
        
        match output {
            Ok(output) if output.status.success() => {
                let stdout = String::from_utf8_lossy(&output.stdout);
                if let Some(status_char) = stdout.chars().next() {
                    match status_char {
                        ' ' => Ok("clean".to_string()),
                        '+' => Ok("modified".to_string()),
                        '-' => Ok("not_initialized".to_string()),
                        'U' => Ok("conflicts".to_string()),
                        _ => Ok("unknown".to_string()),
                    }
                } else {
                    Ok("unknown".to_string())
                }
            }
            _ => Ok("unknown".to_string())
        }
    }
    
    fn get_current_commit(&self, module_path: &PathBuf) -> Result<Option<String>> {
        let output = std::process::Command::new("git")
            .args(&["rev-parse", "HEAD"])
            .current_dir(module_path)
            .output();
        
        match output {
            Ok(output) if output.status.success() => {
                let commit = String::from_utf8_lossy(&output.stdout).trim().to_string();
                if commit.len() >= 8 {
                    Ok(Some(commit[..8].to_string()))
                } else {
                    Ok(Some(commit))
                }
            }
            _ => Ok(None)
        }
    }
    
    async fn update_single_module(
        &self, 
        _module_name: &str, 
        module_info: &SubmoduleInfo, 
        module_path: &PathBuf
    ) -> Result<()> {
        // Fetch latest changes
        println!("{}", "Fetching latest changes...".dimmed());
        let fetch_output = std::process::Command::new("git")
            .args(&["fetch", "origin"])
            .current_dir(module_path)
            .output()?;
        
        if !fetch_output.status.success() {
            return Err(anyhow::anyhow!("Failed to fetch: {}", 
                String::from_utf8_lossy(&fetch_output.stderr)));
        }
        
        // Switch to target branch
        println!("{}", format!("Switching to branch {}...", module_info.branch).dimmed());
        let checkout_output = std::process::Command::new("git")
            .args(&["checkout", &module_info.branch])
            .current_dir(module_path)
            .output()?;
        
        if !checkout_output.status.success() {
            return Err(anyhow::anyhow!("Failed to checkout {}: {}", 
                module_info.branch, String::from_utf8_lossy(&checkout_output.stderr)));
        }
        
        // Pull latest changes
        let pull_output = std::process::Command::new("git")
            .args(&["pull", "origin", &module_info.branch])
            .current_dir(module_path)
            .output()?;
        
        if !pull_output.status.success() {
            return Err(anyhow::anyhow!("Failed to pull: {}", 
                String::from_utf8_lossy(&pull_output.stderr)));
        }
        
        // Stage the submodule update
        let add_output = std::process::Command::new("git")
            .args(&["add", &module_info.path])
            .current_dir(&self.ai_root)
            .output()?;
        
        if !add_output.status.success() {
            return Err(anyhow::anyhow!("Failed to stage submodule: {}", 
                String::from_utf8_lossy(&add_output.stderr)));
        }
        
        Ok(())
    }
    
    async fn auto_commit_changes(&self, updated_modules: &[(String, Option<String>, Option<String>)]) -> Result<()> {
        println!("{}", "💾 Auto-committing changes...".blue());
        
        let mut commit_message = format!("Update submodules\n\n📦 Updated modules: {}\n", updated_modules.len());
        for (module_name, old_commit, new_commit) in updated_modules {
            commit_message.push_str(&format!(
                "- {}: {} → {}\n", 
                module_name,
                old_commit.as_deref().unwrap_or("unknown"),
                new_commit.as_deref().unwrap_or("unknown")
            ));
        }
        commit_message.push_str("\n🤖 Generated with aigpt-rs submodules update");
        
        let commit_output = std::process::Command::new("git")
            .args(&["commit", "-m", &commit_message])
            .current_dir(&self.ai_root)
            .output()?;
        
        if commit_output.status.success() {
            println!("{}", "✅ Changes committed successfully".green());
        } else {
            return Err(anyhow::anyhow!("Failed to commit: {}", 
                String::from_utf8_lossy(&commit_output.stderr)));
        }
        
        Ok(())
    }
}